README.txt
WebAgent is a C based http client, robot & html parser and indexer
(using sql dbms, such as mysql as datastore) that aims to watch, index
and notify subscribers of relevant changes to any web page.

It has run for about 10 years on http://www.webagent.nl/  
(see http://web.archive.org/web/20090603071755/http://www.webagent.nl/?)
serving up to 10000 subscribers, before the hard drive crashed down and 
burned all subscription data for good.

For me the project served as an exercise in C, in which I wanted to 
implement at least:
* a http client (wrote it from scratch) 
* a html parser (wrote it using lex)
This was definitely a sign of the (my) times, and nowadays this
functionality is much more better and stable built in a few lines of
developer code in e.g. python 

The complete WebAgent project as checked into repository consists of this 
http client and html parser, but also a newsletter / mail application 
(to mail out the detected changes) and a complete website (in PHP 3 I 
believe...) for endusers to actually subscribe to their provided urls 
and configure their filters and settings.

Please bare, the PHP code is ancient. The C code is of an amateur. 
Nevertheless, the thing worked pretty stable, was very fast, and served 
a decent sized base of almost 10K subscribers for almost 10 years.


doc/CHANGES
May 10
- fixed bug in wn_fmt_retract that retracted the whole tree when links 
  accidentally matched. there is still rather unlikely bug in this that
  should be solved

May 7
- added a link-follower in wn_fmt_fetch that recursively follows a link
  in the page when it is the _only_ link, i.e. we assume it is a redirect

Nov 3
- finally fixed wnletter to use templates, db library etc.

Oct 24
- fixed another bug in wn_net_http_status causing a HTTP_STAT_OK even when it
  is not ok (causing the rest of the header to be ignored in scanning)
- changed wn_net_scan_header function to not use strdup anymore

Oct 18
- fixed stupid bug in wn_db_email_by_uid causing out of bounds errors if 
  called more than MAX_RES times in a single db_open -> db_free process

Oct 17
- created extra navigation links on site, faq, help, tos, about etc.
- made screenshots of all pages and combined them in an explaining interfacedoc

Oct 16
- added field change_freq to database to keep track of average change freq
- changed wnprune to additionally calculate the change_freq as avg updates
  per week for all documents and store it in the database
- used the change_freq field in the website display.php and menu.php to 
  display average number of emails sent in a week per document
- fixed some minor layout issues in joining process

Oct 15
- moved swish index to its own directory and added a swish.conf configfile
- added wnindex program to filter out most frequently encountered words
  in the swish index
- added swish.sh shell program to generate index, dump it and run wnindex
- used output from wnindex in the website menu.php to offer predefined searches

Oct 10
- fixed wnclean to use the template system
- moved db functions to own lib subdir, to enable other dbms engines
- added db/wrappers directory where uid->email mapping can be made external

Oct 6
- added new column for #subcribers per doc and sorting support in display.php

Sep 27
- added wn_mail_spool and wn_mail_process to lib to enforce a new way of
  handling mail through a spool directory which makes it all more reliable
- changed wnmail and wncontrol to send mail through spool
- added a new child process to wncontrol to process spooled mail

Sep 26
- changed wnclean to use the new db functions
- tested wncontrol with new db functions
- changed db function wn_db_query to reconnect if server passes away
- changed template usage in wncontrol to load template only once

Sep 25
- added wn_db to lib directory to provide core database functions
- changed wncontrol to use the new db functions (not compiled/tested yet)

Sep 23
- added wn_str_escape to escape quotes before using in sql statements
- changed title field in database to be filled automatically from <title>
  header in html pages

Sep 18
- added external indexer swish to provide full-text search on website

Sep 13
- added categories for documents in database
- changed php scripts to include category displays
- changed wnclean to include category delete

Sep 7
- fixed bug in wncontrol that childproc SIGSEGV'd when using templates for mail

Sep 6
- added wn_template in library, it reads and parses templates for use in mail
- changed wn_mail_fmt to make use of templates (in etc/templates)
- changed wnmail to make use of templates (in etc/templates)
- changed wncontrol to make use of templates, not yet proven stable

Sep 5
- added a err_no -> err_str mapping with wn_get_errstr function
- finished wnclean utility, included option to force-delete one document

Aug 30
- added filter field in database used in wncontrol with wn_fmt_typecount to 
  filter priority mails
- added errors table in which wncontrol keeps track of how many times the same 
  error occurred one after another so appropriate action can take place

Aug 29
- added wn_fmt_typecount function for counting elements of specific types 
- added wn_fmt_retract function to undo the latest addtag for use in:
- changed parser to excluse all clickable images (these are mostly ads)

Aug 28
- rewrote wn_net_http* functions to support generic header parsing
- added http/1.1 support for chunking, still behind a #undef for it is buggy
- added wn_str_shift as a safe function instead of ptr = &ptr[shift]
- introduced new SIGSEGV bug when loading large urls :-( 
- resolved minor bug, forgot to initialise url->data as url->data[0] = 0
- changed dynamic allocation trigger in wn_net_fetch to do it sooner and more
- (still SIGSEGV exists, only now it is not reproducable anymore in gdb)
- introduced WN_FMT_CHAR_NULL (a control char) to represent NULL tags
- now using If-Modified-Since header from/to database to significantly
  reduce network load: when applied, it returns WN_SKIP in wn_get_errno

Aug 27
- added more http header parsing in order to determine 30x relocations
- 30x relocations are retrieved and parsed by recursively calling wn_net_fetch

Aug 25
- started logging changes
- resolved major bug, no test on NULL ptr in wn_fmt_load caused SIGSEGV


www/index.php
<? 
	$font_face = "Arial,Geneva";
	$left_headercolor = "#003366";
	$left_bgcolor = "#666699";
	$left_subbgcolor = "#aeb3cc";
	$left_sub2bgcolor = "#cdcdcd";
    	$right_headercolor = "#993300";
	$right_subheadercolor = "#cc9999";
	$right_bgcolor = "#cc6633";
	$right_subbgcolor = "#ffcc9b";
	$text_color = "#ffffff";
	$text2_color = "#000000";
	$text3_color = "#333333";
   	$border_color = "#000000";
	$bf_bgcolor_normal = "#ffee9b";
	$bf_bgcolor_dimmed = "#ddcc9a";
	$padwidth = 7; $padheight = 10;
	$padbetween = 10;
	if (isset ($wn_lang) && $wn_lang)
		$lang = $wn_lang;
	else
		$lang = "nl";
	$max_title_len = 30;
	$title_ext = " ..";

	$banner[0] = "img/banner.gif";    $adurl[0] = "http://www.webagent.nl/";	
	$banner[1] = "img/bannerdds.gif"; $adurl[1] = "http://vacatures.dds.nl/";
	$banneridx = 0;

	srand((double)microtime()*1000000); 
	if (rand(0,4) < 1)
		$banneridx = 1;;

	include "db.php";

	$auth_ok = 0;

	function vpad ($height, $color, $colspan) {
		echo "<tr><td colspan=$colspan bgcolor=$color height=$height><img src=img/trans.gif border=0 height=$height></td></tr>\n";
	}

	function window_start ($title_left, $title_right, $header_textcolor, $left_headercolor, $content_bgcolor) {
		global $padwidth, $padheight;
		global $border_color;
	
		if (!strlen ($title_right))
			$title_right="&nbsp;";

		echo "<table border=0 cellpadding=0 cellspacing=0 width=100%>\n";

		echo "<tr><td colspan=3 bgcolor=$left_headercolor><table border=0 cellpadding=0 cellspacing=0 width=100%><tr><td bgcolor=$left_headercolor nowrap><font face=Arial,Geneva size=-1 color=$header_textcolor><b>&nbsp;$title_left</b></td>\n";
		echo "<td bgcolor=$left_headercolor valign=middle align=right nowrap>" . $title_right . "&nbsp;</td></tr></table></td></tr>\n";

		echo "<tr><td colspan=3 bgcolor=$content_bgcolor height=$padheight><img src=img/trans.gif height=$padheight></td></tr>\n";

		echo "<tr><td width=$padwidth bgcolor=$content_bgcolor><img src=img/trans.gif width=$padwidth border=0></td><td bgcolor=$content_bgcolor valign=top>\n";
	}

	function window_end ($content_bgcolor) {
		global $padheight, $padwidth;
		global $border_color;

		echo "</td><td width=$padwidth bgcolor=$content_bgcolor><img src=img/trans.gif width=$padwidth border=0></td></tr>\n";

		echo "<tr><td colspan=3 bgcolor=$content_bgcolor height=$padheight><img src=img/trans.gif height=$padheight></td></tr>\n";

		echo "</table>\n";
	}

	function page_end () {
		echo "</td></tr></table>\n";
		echo "</td></tr></table>\n";
		echo "</body>\n";
		exit;
	}

	if ($l) {
		SetCookie ("wn_lang", $l, time() +36000000, "/");
		$lang = $l;
	}

	include "index_text.php";

	if ($o) SetCookie ("wn_auth", "", "", "/");

	if (!$o && $wn_auth && (!($e && $pw))) {
		$koek = $wn_auth;
		for ($i = 0; $i < strlen ($koek); $i++)
			$koek[$i] = chr (ord ($koek[$i]) -1);
		$e = strtok ($koek, ":");
		$pw = strtok (":");
	}

	$err = "";
	if ($pw)
		$pw = substr ($pw, 0, 8);

	if (($j == 1) && $pw && $e) {
		if (!eregi ("(.*)@(.*)\.(.*)", $e))
			$err = $txt_err_email_invalid;
		else {
			$res = mysql_query ("select id from users where email = '$e'", $conn);
			if (mysql_num_rows ($res)) 
				$err = $txt_err_email_exists;
			else {
				$now = date ("U");
				$res = mysql_query ("insert into users values (0, '$e','$pw', $now)", $conn);
			}
		}
	}

	if (($j == 2) && $e) {
		$res = mysql_query ("select passwd from users where email = '$e'", $conn);
		if (mysql_num_rows ($res)) {
			mail ($e, $txt_mail_passwd_subject, $txt_mail_passwd_pre . "Email   : $e\nPassword: " . mysql_result ($res, 0, "passwd") . $txt_mail_passwd_post);
			$sent = 1;
		}
	}

	if ($e && $pw && ($err == "")) {
		$res = mysql_query ("select id, lastlogon from users where email = '$e' and passwd = '$pw'", $conn);
		if (mysql_num_rows ($res)) {
			$uid = 0 + mysql_result ($res, 0, "id");
			$lastlogon = 0 + mysql_result ($res, 0, "lastlogon");
			$koek = "$e:$pw";
			for ($i =0; $i < strlen ($koek); $i++) 
				$koek[$i] = chr (ord ($koek[$i]) + 1);
			SetCookie ("wn_auth", $koek, time()+36000000, "/");
			$auth_ok = 1;
		} 
	} 


	if ($auth_ok && $t && $a) {
		switch ($a) {
		case "a":
			$tmpl = 0;
			if ($t > 0) 
				$res = mysql_query ("select id from docs where id = $t");
			else {
				$url = ereg_replace ("http://", "", $url);
				$host = strtok ($url, "/");
				$uri = substr ($url, strlen ($host));
				if (!$uri) $uri = "/";
				if (strlen ($host)) {
					$res = mysql_query ("select id, alt_host, alt_uri from docs_banned where host like '%" . $host . "%' and (uri = '' or uri = '$uri')");
					if (!mysql_num_rows ($res)) {
						$res = mysql_query ("select id from docs where host = '$host' and uri = '$uri'");
						if (!mysql_num_rows ($res)) {
							$url = addslashes ($url);
							$res = mysql_query ("insert into docs (id, host, title, uri, updated, status, changed, changed_serv, watchers, change_freq) values (0, '$host', '$url', '$uri', 0, 0, 0, '', 0, 0)");
		
							$res = mysql_query ("select id from docs where host = '$host' and uri = '$uri'");
						}
					} else {
						$alt_host = mysql_result ($res, 0, "alt_host");	
						$alt_uri = mysql_result ($res, 0, "alt_uri");	
						if ($alt_host && strlen ($alt_host)) {
							if (!$alt_uri) $alt_uri = "/";
							$res = mysql_query ("select id from docs where host = '$alt_host' and uri = '$alt_uri'", $conn);

							if (!mysql_num_rows ($res)) {
								$alt_uri = addslashes ($alt_uri);
								$res = mysql_query ("insert into docs (id, host, title, uri, updated, status, changed, changed_serv, watchers, change_freq) values (0, '$alt_host', 'http://$alt_host$alt_uri', '$alt_uri', 0, 0, 0, '', 0, 0)");
			
								$res = mysql_query ("select id from docs where host = '$alt_host' and uri = '$alt_uri'");
							}
							$tmpl = 2;
						} else $aerr = $txt_doc_banned;
					} 
				} else $aerr = $txt_err_fill_all_fields;
			}

			if (!$aerr) {
				if (mysql_num_rows ($res)) {
					$did = 0 + mysql_result ($res, 0, "id");
					if ($cat_id) {
						$res = mysql_query ("select cid from docs_cats where did = $did");
						if (!mysql_num_rows ($res)) {
							$res = mysql_query ("insert into docs_cats (did, cid) values ($did, $cat_id)", $conn);
							$res = mysql_query ("update cats set size = size + 1 where id = $cat_id", $conn);
						}
					}
					$res = mysql_query ("select uid from watch where uid = $uid and did = $did");
					if (!mysql_num_rows ($res)) {
						$res = mysql_query ("insert into watch (uid, did, ubase, email_prio, keywords, letter_update, filter, title, bfilter, tmpl) values ($uid, $did, 0, 1, '', 0, 3, '', '', $tmpl)");
						$res = mysql_query ("update docs set watchers = watchers + 1 where id = $did");
					}

					$d = $did;
					$a = "s"; 
				} else $aerr = $txt_err_try_again;
			}
			break;
		case "b":
			$bfilter = "";	
			if (is_array ($bf_id)) {
				for ($i = 0; $i < count ($bf_id); $i++) {
					if ($bf_id[$i] < 256)
						$bfilter .= chr($bf_id[$i]);
				}
			}
			mysql_query ("update watch set bfilter = '$bfilter' where did = $t and uid = $uid", $conn);
			break;
		case "s2":
			$filter = 0;
			$filter += $filter_link;
			$filter += $filter_text;
			$filter += $filter_image;
			$keywords = addslashes ($keywords);
			$mail = 0;
			$letter = 0;
			if ($period > 1) 
				$letter = $period;
			else if ($period == 1)
				$mail = 1;
			if (!isset ($title))
				$title = "";
			else
				$title = addslashes ($title);
			$tmpl = 0 + $tmpl;
	
			mysql_query ("update watch set email_prio = $mail, letter_update = $letter, keywords = '$keywords', filter = $filter, title = '$title', tmpl = $tmpl where did = $t and uid = $uid", $conn);
			break;
		case "u2":
			if (isset ($npw) || isset ($npw2)) {
				if ((!($npw == $npw2)) || (!strlen ($npw))) {
					$uerr = $txt_err_passwd_not_same;
					$a = "u";
				} else {
					$res = mysql_query ("update users set passwd = '$npw' where id = $uid", $conn);
				}
			}

			if (isset ($ne)) {
				$res = mysql_query ("update users set email = '$ne' where id = $uid", $conn);
			}

			if (isset ($qt) || isset ($qt2)) {
				if (!($qt && $qt2)) {
					$uerr = $txt_err_quit_not_verified;
					$a = "u";
				} else {
					$res = mysql_query ("select did from watch where uid = $uid", $conn);
					for ($i = 0; $i < mysql_num_rows ($res); $i++) {
						$did = 0 + mysql_result ($res, $i, "did");
						if ($did)
							$res2 = mysql_query ("update docs set watchers = watchers - 1 where id = $did", $conn);
					}
					$res = mysql_query ("delete from watch where uid = $uid", $conn);
					$res = mysql_query ("delete from users where id = $uid", $conn);
				}
			}

			break;
		case "d":
			$res = mysql_query ("delete from watch where uid = $uid and did = $t");
			$res = mysql_query ("update docs set watchers = watchers - 1 where id = $t");
			break;
		case "m0":
			$res = mysql_query ("update watch set email_prio = 0 where uid = $uid and did = $t");
			break;
		case "m1":
			$res = mysql_query ("update watch set email_prio = 1 where uid = $uid and did = $t");
			break;
		case "d0":
			$res = mysql_query ("update watch set letter_update = 0 where uid = $uid and did = $t");
			break;
		case "d1":
			$res = mysql_query ("update watch set letter_update = 24 where uid = $uid and did = $t");
			break;
		case "b1":
			$res = mysql_query ("update watch set letter_update = 48 where uid = $uid and did = $t");
			break;
		case "w1":
			$res = mysql_query ("update watch set letter_update = 168 where uid = $uid and did = $t");
			break;
		case "t1":
			$res = mysql_query ("update watch set title = '$title' where uid = $uid and did = $t");
			break;
		default:
		}
	}

	if ($w) {
		exec ("/usr/local/bin/swish -w '$w' -m 10 -f /usr/local/wn/swish/swish.index", $res);
		for ($i = 0; $i < count ($res); $i++) {
			if (($res[$i][0] != "#") && (!strstr ($res[$i], "err"))) {
				$arr = split (" ", $res[$i]);
				$results[] = $arr[1];
			}
		}
	}

	$topbutton .= "<a href=index.php?d=0&c=0 onMouseOver=\"document.close_right.src = 'img/close_orange_rollover.gif';\" onMouseOut=\"document.close_right.src = 'img/close_orange_normal.gif';\" onMouseDown=\"document.close_right.src = 'img/close_orange_click.gif';\"><img name=close_right src=img/close_orange_normal.gif width=12 height=12 border=0 alt=\"$txt_alt_backtop\"></a>";

	if (strlen ($QUERY_STRING))
		$add = "?$QUERY_STRING";

	echo "<head>\n<title>WebAgent</title>\n";
	echo "<meta http-equiv=Refresh content=\"1800;URL=/$add\">\n";
	if ($lang == "nl") {
		echo "<meta name=\"keyword\" value=\"website updates, veranderingen, nieuwsbrief\">\n";
		echo "<meta name=\"description\" value=\"WebAgent registreert veranderingen in websites en mailt deze direkt of in een dagelijkse nieuwsbrief naar leden. Webmasters kunnen WebAgent als dienst in hun site opnemen.\">\n";
	} else {
		echo "<meta name=\"keywords\" value=\"website updates, update monitor, change detection, automatic newsletter\">\n";
		echo "<meta name=\"description\" value=\"WebAgent archives changes in websites and mails them directly or periodically in a newsletter to subscribers. Webmasters can integrate WebAgent in their sites as a service.\">\n";
	}
	echo "</head>\n<body background=img/bg.gif bgcolor=#999999 link=$text2_color alink=$text2_color vlink=$text2_color>\n";
?>
<table border=0 width=100% cellpadding=0 cellspacing=0>
<tr>
<td colspan=2 width=100% height=6 bgcolor=#336666><img src=img/trans.gif height=6></td>
</tr>
<tr>
<td valign=top width=276 bgcolor=#336666 align=right> 
<?
	echo "<a href=index.php?d=0&c=0><img src=img/logo.gif width=276 height=55 border=0 align=left alt=\"$txt_alt_backtop\"></a><br clear=all><table border=0 width=100% cellpadding=0 cellspacing=0><tr><td nowrap><font face=$font_face size=-4 color=$text_color>&nbsp;&nbsp;";
	if ($lang == "nl") {
		$qs = ereg_replace ("&(l=nl|l=en)", "", $QUERY_STRING);
		echo "<a href=\"index.php?$qs&l=en\"><font color=$text_color>EN</font></a>/";
	} else
		echo "<b>EN</b>/";
	if ($lang == "en") {
		$qs = ereg_replace ("&(l=en|l=nl)", "", $QUERY_STRING);
		echo "<a href=\"index.php?$qs&l=nl\"><font color=$text_color>NL</font></a>";
	} else
		echo "<b>NL</b>";
	echo "<td nowrap align=right><font face=$font_face size=-4 color=$text_color>&nbsp;&nbsp;$txt_intro_short</td></tr></table>\n";
?>
</td><td valign=top width=500 bgcolor=#336666 align=center nowrap>
<img src=img/trans.gif width=12 border=0><a href=<? echo $adurl[$banneridx]; ?>><img src=<? echo $banner[$banneridx] ; ?> width=476 height=60 border=0 valign=middle></a><img src=img/trans.gif width=12 border=0>
</td></tr>
<tr>
<td colspan=2 width=100% height=3 bgcolor=#336666><img src=img/trans.gif height=3></td>
</tr>
</table>

<table border=0 width=100% cellpadding=0 cellspacing=0>
<? 	
	echo "<tr><td colspan=3 height=$padbetween><img src=img/trans.gif height=$padbetween border=0></td></tr>\n";
	echo "<tr><td valign=top height=40> \n";
	if ($auth_ok) {
		include "menu.php";
	} else {
		include "login.php";
	}
	echo "</td><td width=$padbetween><img src=img/trans.gif width=$padbetween border=0></td>\n";
	echo "<td align=left valign=top width=70%><font face=Arial,Geneva size=-1>\n";

	switch ($a) {
	case "s":
		if ($auth_ok)
			include "settings.php";
		break;
	case "u":
		if ($auth_ok)
			include "user.php";
		break;
	case "f":
		include "faq.php";
		break;
	default:
		include "display.php";
	}

	page_end ();
?>
